drivers/clock_control/clock_control_nrf2_auxpll.c

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_auxpll
+
+#include <errno.h>
+#include <stdint.h>
+
+#include <zephyr/arch/cpu.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/nrf_clock_control.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/toolchain.h>
+
+#include <hal/nrf_auxpll.h>
+#include <zephyr/dt-bindings/clock/nrf-auxpll.h>
+
+#if defined(CONFIG_CLOCK_CONTROL_NRF2)
+#include "clock_control_nrf2_common.h"
+#else
+int api_nosys_on_off(const struct device *dev, clock_control_subsys_t sys)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sys);
+
+	return -ENOSYS;
+}
+#endif
+
+/* Check dt-bindings match MDK frequency division definitions*/
+BUILD_ASSERT(
+	NRF_AUXPLL_FREQ_DIV_MIN			== NRF_AUXPLL_FREQUENCY_DIV_MIN		&&
+	NRF_AUXPLL_FREQ_DIV_AUDIO_44K1	== NRF_AUXPLL_FREQUENCY_AUDIO_44K1	&&
+	NRF_AUXPLL_FREQ_DIV_USB24M		== NRF_AUXPLL_FREQUENCY_USB_24M		&&
+	NRF_AUXPLL_FREQ_DIV_AUDIO_48K	== NRF_AUXPLL_FREQUENCY_AUDIO_48K	&&
+	NRF_AUXPLL_FREQ_DIV_MAX			== NRF_AUXPLL_FREQUENCY_DIV_MAX,
+	"Different AUXPLL frequency definitions in MDK and devicetree bindings");
+
+/* maximum lock time in ms, >10x time observed experimentally */
+#define AUXPLL_LOCK_TIME_MAX_MS  20
+/* lock wait step in ms*/
+#define AUXPLL_LOCK_WAIT_STEP_MS 1
+
+struct dev_data_auxpll {
+	struct onoff_manager mgr;
+	onoff_notify_fn notify;
+	const struct device *dev;
+};
+
+struct clock_control_nrf_auxpll_config {
+	NRF_AUXPLL_Type *auxpll;
+	uint32_t ref_clk_hz;
+	uint32_t ficr_ctune;
+	nrf_auxpll_config_t cfg;
+	nrf_auxpll_freq_div_ratio_t frequency;
+	nrf_auxpll_ctrl_outsel_t out_div;
+};
+
+static int clock_control_nrf_auxpll_on(struct dev_data_auxpll *dev_data)
+{
+	const struct clock_control_nrf_auxpll_config *config = dev_data->dev->config;
+	bool locked;
+	unsigned int wait = 0U;
+	nrf_auxpll_task_trigger(config->auxpll, NRF_AUXPLL_TASK_START);
+
+	do {
+		locked = nrf_auxpll_mode_locked_check(config->auxpll);
+		if (!locked) {
+			printk("AUXPLL not locked, waiting...\n");
+			k_msleep(AUXPLL_LOCK_WAIT_STEP_MS);
+			wait += AUXPLL_LOCK_WAIT_STEP_MS;
+		}
+	} while (wait < AUXPLL_LOCK_TIME_MAX_MS && !locked);
```

Comment:
Check out the [`WAIT_FOR()`](https://github.com/zephyrproject-rtos/zephyr/blob/d8b37753061a9e0ace95204eccddee1f4b4c477b/include/zephyr/sys/util.h#L882) macro :)